### PR TITLE
[DeepSpeed] Remove partitioning of model in ZeRO 3

### LINKED ISSUE #10510: DeepSpeed stage 3 and mixed precision cause an error
## 🐛 Bug

Using `strategy="deepspeed_stage_3"` and `precision=16` causes an error

### To Reproduce

```python
import os
import torch
from torch.utils.data import DataLoader, Dataset
from deepspeed.ops.adam import DeepSpeedCPUAdam, FusedAdam
from pytorch_lightning import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return FusedAdam(self.layer.parameters(), lr=0.1)
        # return torch.optim.Adam(self.parameters(),lr = .1)

def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    test_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        enable_model_summary=False,
        logger=False,
        enable_checkpointing=False,
        gpus = 4,
        precision=16,
        strategy = "deepspeed_stage_3"
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
    trainer.test(model, dataloaders=test_data)


if __name__ == "__main__":
    run()

```

I get the following error:

```shell
Traceback (most recent call last):
  File "bug.py", line 69, in <module>
    run()
  File "bug.py", line 64, in run
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
  File "/home/kirill.trapeznikov/miniconda3/envs/semafor_nlg/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 735, in fit
    self._call_and_handle_interrupt(
  File "/home/kirill.trapeznikov/miniconda3/envs/semafor_nlg/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 682, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/home/kirill.trapeznikov/miniconda3/envs/semafor_nlg/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 770, in _fit_impl
    self._run(model, ckpt_path=ckpt_path)
  File "/home/kirill.trapeznikov/miniconda3/envs/semafor_nlg/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1182, in _run
    self._pre_dispatch()
  File "/home/kirill.trapeznikov/miniconda3/envs/semafor_nlg/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1217, in _pre_dispatch
    self.accelerator.pre_dispatch(self)
  File "/home/kirill.trapeznikov/miniconda3/envs/semafor_nlg/lib/python3.8/site-packages/pytorch_lightning/accelerators/accelerator.py", line 136, in pre_dispatch
    self.training_type_plugin.pre_dispatch()
  File "/home/kirill.trapeznikov/miniconda3/envs/semafor_nlg/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/deepspeed.py", line 397, in pre_dispatch
    self.init_deepspeed()
  File "/home/kirill.trapeznikov/miniconda3/envs/semafor_nlg/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/deepspeed.py", line 474, in init_deepspeed
    self._initialize_deepspeed_train(model)
  File "/home/kirill.trapeznikov/miniconda3/envs/semafor_nlg/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/deepspeed.py", line 507, in _initialize_deepspeed_train
    model, deepspeed_optimizer = self._setup_model_and_optimizer(model, optimizer, scheduler)
  File "/home/kirill.trapeznikov/miniconda3/envs/semafor_nlg/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/deepspeed.py", line 431, in _setup_model_and_optimizer
    deepspeed_engine, deepspeed_optimizer, _, _ = deepspeed.initialize(
  File "/home/kirill.trapeznikov/miniconda3/envs/semafor_nlg/lib/python3.8/site-packages/deepspeed/__init__.py", line 131, in initialize
    engine = DeepSpeedEngine(args=args,
  File "/home/kirill.trapeznikov/miniconda3/envs/semafor_nlg/lib/python3.8/site-packages/deepspeed/runtime/engine.py", line 223, in __init__
    self._configure_optimizer(optimizer, model_parameters)
  File "/home/kirill.trapeznikov/miniconda3/envs/semafor_nlg/lib/python3.8/site-packages/deepspeed/runtime/engine.py", line 905, in _configure_optimizer
    self.optimizer = self._configure_zero_optimizer(basic_optimizer)
  File "/home/kirill.trapeznikov/miniconda3/envs/semafor_nlg/lib/python3.8/site-packages/deepspeed/runtime/engine.py", line 1152, in _configure_zero_optimizer
    optimizer = FP16_DeepSpeedZeroOptimizer_Stage3(
  File "/home/kirill.trapeznikov/miniconda3/envs/semafor_nlg/lib/python3.8/site-packages/deepspeed/runtime/zero/stage3.py", line 905, in __init__
    self.create_reduce_and_remove_grad_hooks()
  File "/home/kirill.trapeznikov/miniconda3/envs/semafor_nlg/lib/python3.8/site-packages/deepspeed/runtime/zero/stage3.py", line 1885, in create_reduce_and_remove_grad_hooks
    param.all_gather()
  File "/home/kirill.trapeznikov/miniconda3/envs/semafor_nlg/lib/python3.8/site-packages/deepspeed/runtime/zero/partition_parameters.py", line 590, in all_gather
    return self._all_gather(param_list, async_op=async_op, hierarchy=hierarchy)
  File "/home/kirill.trapeznikov/miniconda3/envs/semafor_nlg/lib/python3.8/site-packages/deepspeed/runtime/zero/partition_parameters.py", line 699, in _all_gather
    ret_value = self._allgather_params_coalesced(all_gather_list, hierarchy)
  File "/home/kirill.trapeznikov/miniconda3/envs/semafor_nlg/lib/python3.8/site-packages/deepspeed/runtime/zero/partition_parameters.py", line 947, in _allgather_params_coalesced
    h = dist._all_gather_base(allgather_params[param_idx],
  File "/home/kirill.trapeznikov/miniconda3/envs/semafor_nlg/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 2070, in _all_gather_base
    work = group._allgather_base(output_tensor, input_tensor)
RuntimeError: output tensor must have the same type as input tensor
```

### Expected behavior

it should work, right?

### Environment

```
* CUDA:
        - GPU:
                - GeForce RTX 2080 Ti
                - GeForce RTX 2080 Ti
                - GeForce RTX 2080 Ti
                - GeForce RTX 2080 Ti
                - GeForce RTX 2080 Ti
                - GeForce RTX 2080 Ti
                - GeForce RTX 2080 Ti
                - GeForce RTX 2080 Ti
        - available:         True
        - version:           11.3
* Packages:
        - numpy:             1.21.1
        - pyTorch_debug:     False
        - pyTorch_version:   1.10.0+cu113
        - pytorch-lightning: 1.5.1
        - tqdm:              4.62.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.11
        - version:           #1 SMP Wed Feb 3 15:06:38 UTC 2021
  
```

- Any other relevant information: `deepspeed   0.5.6`

### LINKED ISSUE #10640: Remove optimizer for DeepSpeed inference
## Refactor/Feature

Recently in DeepSpeed it was merged to support making the optimizer optional for inference https://github.com/microsoft/DeepSpeed/pull/1514.

Currently in our plugin we pass in an optimizer always: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/training_type/deepspeed.py#L573. This is required to partition the parameters appropriately, however this shouldn't be needed now :)

We should wait till a release is made for DeepSpeed. Also we need to be cautious if the user uses a later version of deepspeed and does require an optimizer to be passed. @carmocca do you think we need a deprecation process for this?

### Motivation

Cleaner code, and probably less hacky.

cc @borda @justusschock @awaelchli @akihironitta @SeanNaren

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Partially Fixes #10510

Also adds parametrization to a few tests since it's supported, and handles deprecated args with the latest DeepSpeed.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @justusschock @awaelchli @akihironitta @SeanNaren

### CODE REVIEW

**Review by rohitgr7 — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -46,7 +46,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Moved ownership of the `PrecisionPlugin` into `TrainingTypePlugin` and updated all references ([#10570](https://github.com/PyTorchLightning/pytorch-lightning/pull/10570))
 
 
--
+- DeepSpeed does not require optimizer for inference anymore ([#10655](https://github.com/PyTorchLightning/pytorch-lightning/pull/10655))
```
Comment (SeanNaren) [inline CHANGELOG.md:49]:
```suggestion
- DeepSpeed does not require optimizer for inference anymore and does not require lightning module zero 3 partitioning ([#10655](https://github.com/PyTorchLightning/pytorch-lightning/pull/10655))
```

**Review by awaelchli — APPROVED** [review]

**Comment by SeanNaren** [comment]
Seems this is not so simple. If you load the model in inference mode you cannot save (which makes sense as it's the optimizers responsibility to save the ZeRO sharded weights).

Do we think that the case where the user does not call fit before saving should be supported?

```python
trianer = Trainer(...)
trainer.test(model)
trainer.save_checkpoint(...)
```

If so I think it's better we keep the optimizer code as is and use the `configure_optimizers` to define the optimizer again.

cc @carmocca @awaelchli

**Comment by rohitgr7** [comment]
> Seems this is not so simple. If you load the model in inference mode you cannot save (which makes sense as it's the optimizers responsibility to save the ZeRO sharded weights).
> 
> Do we think that the case where the user does not call fit before saving should be supported?
> 
> ```python
> trianer = Trainer(...)
> trainer.test(model)
> trainer.save_checkpoint(...)
> ```
> 
> If so I think it's better we keep the optimizer code as is and use the `configure_optimizers` to define the optimizer again.
> 
> cc @carmocca @awaelchli @rohitgr7 
> 

Even if we do that, we do not force users to define configure_optimizers during inference so possibility of not getting an optimizer still exists. So there are 2 cases now:

1. User calls .fit and then calls .test, which will work right? Since optimizers will be attached during .fit call.
2. User calls .test directly, then I don't think they will do a save_checkpoint because it it's a completely new randomly initialised model then checkpointing is not necessary and if they are using some pretrained model for testing then they will already have the saved checkpoint and calling another won't make sense. 

Maybe you can add an error/warning for the case with no optimizer and saving under deepspeed config.

File: .azure-pipelines/gpu-tests.yml [hunk 2]
```diff
@@ -51,7 +51,7 @@ jobs:
     - bash: |
         python -c "fname = 'requirements/extra.txt' ; lines = [line for line in open(fname).readlines() if 'horovod' not in line] ; open(fname, 'w').writelines(lines)"
         pip install fairscale==0.4.0
-        pip install deepspeed==0.5.4
+        pip install deepspeed==0.5.7
```
Comment (carmocca) [inline .azure-pipelines/gpu-tests.yml:54]: Update this too

https://github.com/PyTorchLightning/pytorch-lightning/blob/8ea39d2c8f68cc33273c3431a310a262e2240cf9/dockers/base-cuda/Dockerfile#L115

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 3]
```diff
@@ -260,12 +259,6 @@ def __init__(
             load_full_weights: True when loading a single checkpoint file containing the model state dict
                 when using ZeRO Stage 3. This differs from the DeepSpeed checkpoint which contains shards
                 per worker.
-
-            partition_module: When True, partitions the ``LightningModule`` across devices when using ZeRO Stage 3.
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/deepspeed.py:264]: I am curious, why removing `partition_module`? It still seems useful in training mode.

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 3 above)
Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/deepspeed.py:264]: Check the related issue: https://github.com/PyTorchLightning/pytorch-lightning/issues/10510

**Comment by SeanNaren** [comment]
Pulled out the other issue, will make a separate PR for the deepspeed inference/optimizer fix.

**Review by awaelchli — APPROVED** [review]

**Comment by SeanNaren** [comment]
Closing this PR till we figure out a fix. I think changes upstream in DeepSpeed need to be made to handle this case: https://github.com/PyTorchLightning/pytorch-lightning/issues/10510#issuecomment-975513569

I'll be making a hotfix that will fix most of the use cases where either the user does not use `configure_sharded_model` or defines the entire model in `configure_sharded_model`.

**Comment by SeanNaren** [comment]
This PR will partially fix the related issue, but there is still the case of partially defined partitioned weights to tackle.

**Pull request merged**